pkg/security/admission/admission.go

```diff
@@ -0,0 +1,318 @@
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string, 0)
+		}
+		pod.Annotations[createPodAnnotationKey(&container)] = providerName
+	}
+
+	return nil
+}
+
+// getSecurityContextForContainer iterates through the providers and generates then validates
+// the sc against the scc.  It operates on a copy of the pod so that it does not change any
+// actual values on the containers since the scc provider will set defaults for anything that isn't
+// specified.
+func (c *constraint) getSecurityContextForContainer(pod *kapi.Pod, containerIndex int, providers map[string]scc.SecurityContextConstraintsProvider) (*kapi.SecurityContext, string, error) {
+	for providerName, provider := range providers {
+		// Create a security context for this provider
+
+		// Since the provider changes values on the container we need to work with a copy so that
+		// as we iterate through the providers we don't get defaults set from previous providers.
+		podCopy, err := createPodCopy(pod)
+		if err != nil {
+			return nil, "", err
+		}
+
+		container := &podCopy.Spec.Containers[containerIndex]
+		context, err := provider.CreateSecurityContext(podCopy, container)
+		if err != nil {
+			return nil, "", err
+		}
+
+		// Validate the copy of the container against the security context provider
+		container.SecurityContext = context
+		errs := provider.ValidateSecurityContext(pod, container)
+		if len(errs) == 0 {
+			return context, providerName, nil
+		}
+	}
+
+	// If we have reached this code, we couldn't find an SCC that matched
+	// the requested security context for this container
+	return nil, "", fmt.Errorf("unable to find a valid security context constraint for container %s", pod.Spec.Containers[containerIndex].Name)
+}
+
+// createPodCopy creates a copy of a pod.
+func createPodCopy(pod *kapi.Pod) (*kapi.Pod, error) {
+	pCopy, err := kapi.Scheme.Copy(pod)
+	if err != nil {
+		return nil, err
+	}
+	podCopy, ok := pCopy.(*kapi.Pod)
+	if !ok {
+		return nil, fmt.Errorf("error converting copied container to a container type")
+	}
+	return podCopy, nil
+}
+
+// createPodAnnotationKey creates a container suffixed annotation key based on validatedSCCAnnotation.
+func createPodAnnotationKey(container *kapi.Container) string {
+	return fmt.Sprintf("%s-%s", allocator.ValidatedSCCAnnotation, container.Name)
+}
+
+// getMatchingSecurityContextConstraints returns constraints from the store that match the group,
+// uid, or user of the service account.
+func getMatchingSecurityContextConstraints(store cache.Store, userInfo user.Info) ([]*kapi.SecurityContextConstraints, error) {
+	matchedConstraints := make([]*kapi.SecurityContextConstraints, 0)
+
+	for _, c := range store.List() {
+		constraint, ok := c.(*kapi.SecurityContextConstraints)
+		if !ok {
+			return nil, errors.NewInternalError(fmt.Errorf("error converting object from store to a security context constraint: %v", c))
+		}
+		for _, userGroup := range userInfo.GetGroups() {
```

Comment:
refactor these two loops into a `contraintAppliesTo(constraint, userInfo) bool` that can return true early on a match